addons/pkg-maven/common/src/main/java/org/commonjava/indy/pkg/maven/content/MavenMetadataGenerator.java

```diff
@@ -371,67 +375,102 @@ public Transfer generateGroupFileContent( final Group group, final List<Artifact
             eventMetadata.set( GROUP_METADATA_EXISTS, true );
             return target;
         }
-        try
-        {
-            target.lockWrite();
 
-            final Metadata md = generateGroupMetadata( group, members, path );
-            if ( md != null )
+        final boolean locked = mergerLock.tryLock();
+
+        if ( locked )
+        {
+            try
             {
-                final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-                try
+                logger.debug( "Start metadata generation for the metadata file for this path {} in group {}", path,
+                              group );
+                final Metadata md = generateGroupMetadata( group, members, path );
```

Comment:
This generateGroupMetadata can be very slow. It includes many slow actions, downloading missing metas, going to Brew to query versions, generating missing metas. It can take quite a few minutes in the worst cases. If you use a global lock for all users, when one user gets the lock and starts working, all the other users have to wait and may get timed out in tryLock(..) and return null, regardless what different paths they are to generate.

Reply:
The timeout is needed, because we don't want to block all other threads, especially when that merging process is slow. Maybe we can set a longer timeout here. I'll set it to 30s.

Reply:
I meant the mergeLock is global not per path. This is very expensive. E.g., if one user does curl http://localhost:8080/api/group/DA/com/sun/xml/bind/jaxb-impl/maven-metadata.xml and it takes 3 minutes. In this 3m, all other meta generations are blocked for 30s (e.g., foo/bar/maven-metadata.xml) and eventually they get null. The old code use target.lockWrite(), which is per path. Does that have a problem?

Reply:
Fixed